### PR TITLE
expression: fix greatest and least function with collation get wrong result

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31789

when deriving collation from expression, greatest and least function are missed, then they will use the `connection_collation` as the collation.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
fix greatest and least function with collation get wrong result
```

### CODE REVIEW

**Review by Defined2014 — APPROVED** [review]

**Review by bb7133 — COMMENTED** [review]
Please add a test case from #31789

**Review by tangenta — APPROVED** [review]
It is better to add an integration test.

File: cmd/explaintest/t/common_collation.test [hunk 1]
```diff
@@ -6,4 +6,21 @@ insert into t values ('啊', '撒旦');
 select coercibility(concat(a, b)) from t;
 select coercibility(convert(concat(a, b) using utf8mb4) collate utf8mb4_general_ci) from t;
 select coercibility(convert('a' using utf8mb4));
-select coercibility(convert('a' using utf8mb4) collate utf8mb4_general_ci);
\ No newline at end of file
+select coercibility(convert('a' using utf8mb4) collate utf8mb4_general_ci);
+
+DROP TABLE IF EXISTS t2;
```
Comment (wjhuang2016) [inline cmd/explaintest/t/common_collation.test:11]: Please add comments for this test case.

**Comment by hawkingrei** [comment]
Please add comments for this test case.

File: cmd/explaintest/t/common_collation.test (hunk 1 above)
Comment (xiongjiwei) [reply cmd/explaintest/t/common_collation.test:12]: Done

**Pull request merged**